### PR TITLE
Refactor and harden size and stats functions 

### LINKED ISSUE #2871: hypertable_size() returns NULL for empty hypertable
**Relevant system information:**
 - TimescaleDB version (output of `\dx` in `psql`): at least 2.0.0

**Describe the bug**
When calling `hypertable_size` on a hypertable that has no rows, the result is `NULL` whereas one might reasonably expect it to return `0`.

**To Reproduce**
1. Create a hypertable, e.g., `foo`
2. Call `hypertable_size` on it before any rows are inserted:
```
SELECT * FROM hypertable_size('bar');
 hypertable_size
-----------------

(1 row)
```

**Expected behavior**
A single row with 0 for the size.

**Actual behavior**
A single row with NULL returned.

### PR DESCRIPTION
**Pull request by erimatnor**
Fix a number of issues with size and stats functions:

* Return `0` size instead of `NULL` in several functions when
  hypertables have no chunks (e.g., `hypertable_size`,
  `hypertable_detailed_size`).
* Return `NULL` when functions are called on non-hypertables instead
  of simply failing with generic error `query returned no rows`.
* Include size of "root" hypertable, which can have non-zero size
  indexes and other objects even if the root table holds no data.
* Make `hypertable_detailed_size` include one additional row for
  storage size of objects on the access node. While the access node
  stores no data, the empty hypertable may still take up some disk
  space.
* Improve test coverage for all size utility functions. In particular,
  add tests on regular tables as well as empty and compressed
  hypertables.
* Several size utility functions that were defined as `PL/pgSQL`
  functions have been converted to simple `SQL` functions since they
  ran only a single SQL query.

The `dist_util` test is moved to the solo test group because,
otherwise, it gives different size output when run in parallel vs. in
isolation.

Fixes #2871

### CODE REVIEW

**Comment by erimatnor** [comment]
Tests are failing because the size for a hypertable differs depending on how the test is run (and maybe platform). In particular, one test is causing issues: Testing the size functions on an hypertable with an empty chunk (which is different from an empty hypertable with no chunks). The tests executes by first inserting a row into the hypertable followed by deleting all tuples. This results in an empty hypertable with a single chunk.

Now, the test has different outcome (in terms of number of bytes reported by size functions), if

* the test is run in isolation, vs.
* the test is run as part of the full regression suite (in parallel with other tests)

So, it seems something is affecting the on-disk size when other tests run in the same database instance.

**Review by svenklemm — CHANGES_REQUESTED** [review]
I think this should include sizes of the parent hypertable as well. While the parent hypertable can not contain data parent metadata (e.g. indexes) can still consume space. 
```
pr2989=# select create_hypertable('t','time');
 create_hypertable
-------------------
 (1,public,t,t)
(1 row)

pr2989=# select pg_indexes_size('t');
 pg_indexes_size
-----------------
            8192
(1 row)

pr2989=# select * from hypertable_detailed_size('t');
 table_bytes | index_bytes | toast_bytes | total_bytes | node_name
-------------+-------------+-------------+-------------+-----------
           0 |           0 |           0 |           0 |
(1 row)

```

**Comment by erimatnor** [comment]
> I think this should include sizes of the parent hypertable as well. While the parent hypertable can not contain data parent metadata (e.g. indexes) can still consume space.
> 
> ```
> pr2989=# select create_hypertable('t','time');
>  create_hypertable
> -------------------
>  (1,public,t,t)
> (1 row)
> 
> pr2989=# select pg_indexes_size('t');
>  pg_indexes_size
> -----------------
>             8192
> (1 row)
> 
> pr2989=# select * from hypertable_detailed_size('t');
>  table_bytes | index_bytes | toast_bytes | total_bytes | node_name
> -------------+-------------+-------------+-------------+-----------
>            0 |           0 |           0 |           0 |
> (1 row)
> ```

I think I agree, and I did think about this case, although I refrained from changing the semantics of the function beyond fixing the NULL issue. Happy to include the root table if people think that makes sense.

File: sql/size_utils.sql [hunk 1]
```diff
@@ -67,46 +67,25 @@ RETURNS TABLE (
 AS '@MODULE_PATHNAME@', 'ts_dist_remote_chunk_info' LANGUAGE C VOLATILE STRICT;
 
 CREATE OR REPLACE FUNCTION _timescaledb_internal.hypertable_local_size(
-    schema_name_in name,
-    table_name_in name)
+	schema_name_in name,
+	table_name_in name)
```
Comment (mkindahl) [inline sql/size_utils.sql:71]: Why do we use schema + table name here but regclass in `hypertable_detailed_size`? Shouldn't we stick to the same conventions in both tables?

**Comment by erimatnor** [comment]
> > I think this should include sizes of the parent hypertable as well. While the parent hypertable can not contain data parent metadata (e.g. indexes) can still consume space.
> > ```
> > pr2989=# select create_hypertable('t','time');
> >  create_hypertable
> > -------------------
> >  (1,public,t,t)
> > (1 row)
> > 
> > pr2989=# select pg_indexes_size('t');
> >  pg_indexes_size
> > -----------------
> >             8192
> > (1 row)
> > 
> > pr2989=# select * from hypertable_detailed_size('t');
> >  table_bytes | index_bytes | toast_bytes | total_bytes | node_name
> > -------------+-------------+-------------+-------------+-----------
> >            0 |           0 |           0 |           0 |
> > (1 row)
> > ```
> 
> I think I agree, and I did think about this case, although I refrained from changing the semantics of the function beyond fixing the NULL issue. Happy to include the root table if people think that makes sense.

Thinking about this more, it has larger implications for distributed hypertables. For instance, we probably need to include the access node in `hypertable_detailed_size` as well as the aggregated value in `hypertable_size`.

File: sql/size_utils.sql [hunk 2]
```diff
@@ -67,46 +67,25 @@ RETURNS TABLE (
...
-               WHERE
-                  schema_name = schema_name_in 
-                  AND table_name = table_name_in 
-               GROUP BY
-                  hypertable_id 
-            ) sub1 
-      ) sub2;
-END;
+	SELECT
+		(COALESCE(sum(ch.total_bytes), 0) - COALESCE(sum(ch.index_bytes), 0) - COALESCE(sum(ch.toast_bytes), 0) + COALESCE(sum(ch.compressed_heap_size), 0))::bigint AS heap_bytes,
```
Comment (mkindahl) [inline sql/size_utils.sql:80]: Nit: you capitalize `COALESCE` but not `sum`. Shouldn't we handle these the same?

File: sql/size_utils.sql [hunk 3]
```diff
@@ -162,50 +138,52 @@ $BODY$;
...
         WHERE c.OID = hypertable;
 
+		IF table_name IS NULL THEN
+		    RAISE EXCEPTION 'relation "%" is not a hypertable', hypertable::text
+			      USING HINT = 'Use pg_relation_size() to get the size of other relations.',
+				  ERRCODE = 'invalid_parameter_value';
+		END IF;
+
         CASE WHEN is_distributed THEN
-            RETURN QUERY SELECT * FROM _timescaledb_internal.hypertable_remote_size(schema_name, table_name);
```
Comment (mkindahl) [inline sql/size_utils.sql:188]: Suggest to return the size of the access node hypertable with `NULL` as the name, similar to how it's done for the non-distributed case.

**Review by mkindahl — COMMENTED** [review]
Looks good, but a few things that I think we should consider:

- Including the size of the main hypertable together with the size of the chunks.
- Including the size of the access node tables with the other data.
- Doesn't it make more sense to return NULL rather than raise an error? It allows queries like `SELECT hypertable_size(oid) FROM pg_class` to work as expected and will return NULL for non-hypertables.

File: sql/size_utils.sql (hunk 1 above)
Comment (erimatnor) [reply sql/size_utils.sql:71]: This is an internal function and since we don't have regclass in our internal metadata, this is what our internal functions work with. Otherwise, I think a larger refactoring requires a larger scope than the issue this PR is trying to fix.

File: sql/size_utils.sql (hunk 2 above)
Comment (erimatnor) [reply sql/size_utils.sql:80]: This was the way it was before these changes. The code has larger issues with formatting and inconsistencies. There was an agreement to fix this is a separate PR since it requires running the code through a SQL formatting tool.

File: sql/size_utils.sql (hunk 3 above)
Comment (erimatnor) [reply sql/size_utils.sql:188]: I plan to do that, although I am not sure returning a NULL name is the best approach. We could just return "access node", although that has other issues if you, e.g., give a data node that name as well (unlikely).

**Comment by erimatnor** [comment]
> Looks good, but a few things that I think we should consider:
>     * Doesn't it make more sense to return NULL rather than raise an error? It allows queries like `SELECT hypertable_size(oid) FROM pg_class` to work as expected and will return NULL for non-hypertables.

Maybe, but I am not sure... Should this apply to any value of the argument or only valid table relations? What if the OID is an index or other type. Or you input a numerical OID that doesn't map to any table? Maybe we return the value of `pg_total_relation_size` for regular tables instead?

I think we'd probably encourage people to do `SELECT hypertable_size(hypertable_name) FROM timescaledb_information.hypertables` instead.

File: sql/size_utils.sql (hunk 3 above)
Comment (mkindahl) [reply sql/size_utils.sql:188]: Well... my thinking was that since you use NULL for the local query, it makes sense to use NULL for the local part of the remote query. If you use a concrete name for the local part in the remote query, you should probably do the same for the fully local query, for consistency.

**Comment by mkindahl** [comment]
> > Looks good, but a few things that I think we should consider:
> > * Doesn't it make more sense to return NULL rather than raise an error? It allows queries like `SELECT hypertable_size(oid) FROM pg_class` to work as expected and will return NULL for non-hypertables.
> 
> Maybe, but I am not sure... Should this apply to any value of the argument or only valid table relations? What if the OID is an index or other type. Or you input a numerical OID that doesn't map to any table? Maybe we return the value of `pg_total_relation_size` for regular tables instead?

In this case, we have OIDs for tables, indexes, and hypertables, and hypertables is a subset of tables. So, what does PG return for tables that expect a table and, for example, does not make sense for an index (for example)? Note that for `pg_table_size`, PostgreSQL returns `NULL` and does not give an error:

```
mats=# select pg_table_size( 'bigint_dst');
 pg_table_size 
---------------
          8192
(1 row)

mats=# select pg_table_size(0);
 pg_table_size 
---------------
        [NULL]
(1 row)
```

To be in line with how PG behaves, it makes more sense to return NULL rather than raise an error.

> I think we'd probably encourage people to do `SELECT hypertable_size(hypertable_name) FROM timescaledb_information.hypertables` instead.

File: sql/size_utils.sql (hunk 3 above)
Comment (mkindahl) [reply sql/size_utils.sql:188]: Note also that I think it makes sense to use `NULL` for the name since the access node does not have a name.

**Comment by k-rus** [comment]
@erimatnor Can you added it into the change log?

**Comment by erimatnor** [comment]
Updated the PR to implement the behavior listed as part of the conclusions at the end of the referenced issue.

File: test/sql/size_utils.sql [hunk 4]
```diff
@@ -164,11 +164,49 @@ SELECT * FROM approximate_row_count();
 SELECT * FROM approximate_row_count(NULL);
 \set ON_ERROR_STOP 1
 
-SELECT * FROM chunks_detailed_size(NULL);
-SELECT * FROM hypertable_detailed_size(NULL);
+-- Test size functions with NULL input
+SELECT * FROM hypertable_size(NULL);
```
Comment (svenklemm) [inline test/sql/size_utils.sql:168]: can you add a test for hypertable_size(0) and hypertable_size(1)

**Review by svenklemm — APPROVED** [review]

**Review by mkindahl — APPROVED** [review]

**Comment by erimatnor** [comment]
@mkindahl @svenklemm I realized the `hypertable_index_size` also didn't include the size of the index on the root hypertable, nor the size of indexes on the access node. I added one more commit with the changes to include the root index size. The commit will be squashed before merge.

File: test/sql/size_utils.sql (hunk 4 above)
Comment (mkindahl) [reply test/sql/size_utils.sql:168]: Did you add tests for this? I don't see them.

**Review by mkindahl — APPROVED** [review]

File: test/sql/size_utils.sql (hunk 4 above)
Comment (erimatnor) [reply test/sql/size_utils.sql:183]: I missed it, but added now.

File: sql/size_utils.sql [hunk 5]
```diff
@@ -67,46 +67,25 @@ RETURNS TABLE (
...
-    table_bytes bigint,
-    index_bytes bigint,
-    toast_bytes bigint,
-    total_bytes bigint)
-LANGUAGE PLPGSQL VOLATILE STRICT AS
+	table_bytes bigint,
+	index_bytes bigint,
+	toast_bytes bigint,
+	total_bytes bigint)
+LANGUAGE SQL VOLATILE STRICT AS
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/size_utils.sql:77]: done.

File: sql/size_utils.sql (hunk 3 above)
Comment (erimatnor) [reply sql/size_utils.sql:188]: Now returns NULL.
